File: jump2.py
#Time Complexity -- O(n)
#Space Complexity -- O(1)

class Solution:
    
#     #Explore every possible path
#     def dfs(self, nums, index, jumps):
#         if index >= len(nums)-1:
#             self.minjumps = min(self.minjumps, jumps)
#             return
        
#         for i in range(1, nums[index]+1):
#             self.dfs(nums, index+i, jumps+1)

    
    
    def jump(self, nums: List[int]) -> int:
        
        
        if not nums or len(nums)<2:
            return 0
        
        n = len(nums)
        currInd = nums[0]
        nextInd = nums[0]
        jumps = 1
        
        for i in range(1, n):
            nextInd = max(nextInd, nums[i]+i)
            if i < n-1 and currInd == i:
                currInd = nextInd
                jumps += 1
                
        
        return jumps
                
            
        
        
        
#         if not nums:
#             return
        
#         self.minjumps = float("inf")
#         self.dfs(nums, 0, 0)
#         return self.minjumps
        

File: jump2.cpp
// Time Complexity : O(n)
// Space Complexity : O(1)
// Any problem you faced while coding this :
//Understand the inner for loop conditition that "i<size-1"
// Your code here along with comments explaining your approach
//1. Create a curr and next interval. 
//2. Traverse through current interval and find the next greatest interval.
//3. Move to the next interval and increment the jump count.

class Solution {
public:
    int jump(vector<int>& nums) {
        //edge
        if(nums.size()==0 || nums.size()<2)
            return 0;
        
        //logic
        int currInterval = nums[0];
        int NextInterval = currInterval;
        int size = nums.size();
        int i =1; int jumps=1;
        while(i<size){
            NextInterval = max(NextInterval, i+nums[i]);
            if(i<size-1 && i == currInterval){
                currInterval = NextInterval;
                jumps++;
            }
            i++;
        }
        return jumps;
    }
};
